crates/sui-types/src/passkey_authenticator.rs

```diff
@@ -0,0 +1,309 @@
+use shared_crypto::intent::{IntentMessage, INTENT_PREFIX_LENGTH};
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::sync::Arc;
+
+#[cfg(test)]
+#[path = "unit_tests/passkey_authenticator_test.rs"]
+mod passkey_authenticator_test;
+
+/// An passkey authenticator with parsed fields. See field defition below. Can be initialized from [struct RawPasskeyAuthenticator].
+#[derive(Debug, Clone, JsonSchema)]
+pub struct PasskeyAuthenticator {
+    /// `authenticatorData` is a bytearray that encodes
+    /// [Authenticator Data](https://www.w3.org/TR/webauthn-2/#sctn-authenticator-data)
+    /// structure returned by the authenticator attestation
+    /// response as is.
+    authenticator_data: Vec<u8>,
+
+    /// `clientDataJSON` contains a JSON-compatible
+    /// UTF-8 encoded serialization of the client
+    /// data which is passed to the authenticator by
+    ///  the client during the authentication request
+    /// (see [CollectedClientData](https://www.w3.org/TR/webauthn-2/#dictdef-collectedclientdata))
+    client_data_json: Vec<u8>,
+
+    /// Normalized r1 signature returned by passkey.
+    /// Initialized from `user_signature` in `RawPasskeyAuthenticator`.
+    #[serde(skip)]
+    signature: Secp256r1Signature,
+
+    /// Compact r1 public key upon passkey creation.
+    /// Initialized from `user_signature` in `RawPasskeyAuthenticator`.
+    #[serde(skip)]
+    pk: Secp256r1PublicKey,
+
+    /// Valid intent parsed from the first 3 bytes of `client_data_json.challenge`.
+    #[serde(skip)]
+    intent: Intent,
+
+    /// Valid tx_digest parsed from the last 32 bytes of `client_data_json.challenge`.
+    tx_digest: [u8; DefaultHash::OUTPUT_SIZE],
+
+    /// Initialization of bytes for passkey in serialized form.
+    #[serde(skip)]
+    bytes: OnceCell<Vec<u8>>,
+}
+
+/// An raw passkey authenticator struct used during deserialization. Can be converted to [struct PasskeyAuthenticator].
+#[derive(Serialize, Deserialize, Debug)]
+pub struct RawPasskeyAuthenticator {
+    pub authenticator_data: Vec<u8>,
+    pub client_data_json: Vec<u8>,
+    pub user_signature: Signature,
+}
+
+/// Convert [struct RawPasskeyAuthenticator] to [struct PasskeyAuthenticator] with validations.
+impl TryFrom<RawPasskeyAuthenticator> for PasskeyAuthenticator {
+    type Error = SuiError;
+
+    fn try_from(raw: RawPasskeyAuthenticator) -> Result<Self, Self::Error> {
+        let client_data_json_parsed: CollectedClientData =
+            serde_json::from_slice(&raw.client_data_json).map_err(|_| {
+                SuiError::InvalidSignature {
+                    error: "Invalid client data json".to_string(),
+                }
+            })?;
+
+        if client_data_json_parsed.ty != ClientDataType::Get {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid client data type".to_string(),
+            });
+        };
+
+        let parsed_challenge = Base64UrlUnpadded::decode_vec(&client_data_json_parsed.challenge)
```

Comment:
Do we know for sure if this needs to be Padded or Unpadded?

Reply:
Looking at the spec it is indeed correct as written as needing to be unpadded: https://w3c.github.io/webauthn/#base64url-encoding